server/runtime/src/main/java/org/infinispan/server/security/realm/CaffeineRealmIdentityCache.java

```diff
@@ -0,0 +1,97 @@
+package org.infinispan.server.security.realm;
+
+import static org.wildfly.common.Assert.checkMinimumParameter;
+
+import java.security.Principal;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.wildfly.security.auth.server.RealmIdentity;
+import org.wildfly.security.cache.RealmIdentityCache;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+
+public final class CaffeineRealmIdentityCache implements RealmIdentityCache {
+
+   private final Cache<Principal, RealmIdentity> identityCache;
+   private final Map<Principal, Set<Principal>> domainPrincipalMap;
+   private final ReadWriteLock lock = new ReentrantReadWriteLock();
+
+   public CaffeineRealmIdentityCache(int maxEntries, long maxAge) {
+      checkMinimumParameter("maxEntries", 1, maxEntries);
+      checkMinimumParameter("maxAge", -1, maxAge);
+      identityCache = Caffeine.newBuilder().maximumSize(maxEntries).expireAfterWrite(maxAge, TimeUnit.MILLISECONDS).build();
```

Comment:
We need to add a listener so that when the entry is removed via expiration it also removes the domainPrincipalMap entry as well.